### PR TITLE
feat - new sample sheet validation

### LINKED ISSUE #2922: Regenerate all v2 sample sheets from before 25-01-24
## Description
The [patch in cg](https://github.com/Clinical-Genomics/cg/pull/2867) that fixed the sample sheet generation did not update all existing sample sheets in Housekeeper, so they need to be generated again.

## Suggested solution

Generate a script that replaces sample sheets in Housekeeper with updated versions

## This can be closed when

~~The script has been run and all sample sheets updated~~
UPDATE: When a new validation has been implemented so that new sample sheets are created.
## Blocked by

*N/A*

### PR DESCRIPTION
**Pull request by diitaz93**
## Description
Fix #2922: Create a more strict validation for v2 sample sheets.

This PR implements a new validation for v2 sample sheets that takes into account 5 aspects:
1. All sections are present
    1. `[Header]`
    2. `[Reads]`
    3. `[BCLConvert_Settings]`
    4. `[BCLConvert_Data]`
1. `IndexSettings` is present in the `[Header]` of the sample sheet
2. Run read and index cycles are present in the `[Reads]` section and are valid
3. The `[BCLConvert_Data]` section has the correct columns (sample validation)
4. The override cycle values are correct. This implies that the cycle values for the reads and the index match with the run cycles specified in the `[Reads]` section and the index2 cycles are in the correct format according to the `IndexSettings` (reverse or forward).

### Added

- A new class `SampleSheetValidator` with the endpoint function `validate_sample_sheet` which will be the new function to validate sample sheets.
- A new class `OverrideCyclesValidator` with the endpoint function `validate_sample` which will validate if the override cycles for a single sample is correct. It is called for each sample inside the `SampleSheetValidator`.
- Tests for new functions

### Changed

- The CLI command `validate_sample_sheet` (`cg demultiplex samplesheet validate`) uses now the new validation.

### Fixed

- Tests for CLI command, parametrised with new fixtures

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

**Review by karlnyr — COMMENTED** [review]
This looks good! Seems like most cases are covered in the validation. However, I think that there are a few pints that may need another look.

File: cg/apps/demultiplex/sample_sheet/validator.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    SampleSheetBcl2FastqSections,
+    SampleSheetBCLConvertSections,
+)
+from cg.exc import OverrideCyclesError, SampleSheetError
+from cg.io.controller import ReadFile
+
+LOG = logging.getLogger(__name__)
+
+FORWARD_INDEX_CYCLE_PATTERN: str = r"I(\d+)N(\d+)"
+REVERSE_INDEX_CYCLE_PATTERN: str = r"N(\d+)I(\d+)"
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/validator.py]: Turned into constants in the constants/demultiplex module

File: cg/apps/demultiplex/sample_sheet/validator.py [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+        if (
+            self.run_index1_cycles == len(self.sample["Index"])
+            and index1_cycle == f"I{self.run_index1_cycles}"
+        ):
+            return
+        if self.is_index_cycle_value_following_pattern(
+            pattern=FORWARD_INDEX_CYCLE_PATTERN,
+            index_cycle=index1_cycle,
+            run_cycles=self.run_index1_cycles,
+            index_sequence=self.sample["Index"],
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/validator.py]: fixed! Replaced all column names for constants

File: cg/apps/demultiplex/sample_sheet/validator.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        Raises:
+            OverrideCyclesError if the index 2 cycle is not valid.
+        """
+        if not self.run_index2_cycles and len(self.sample_cycles) == 3:
+            return
+        index2_cycle: str = self.sample_cycles[2]
+        if not self.sample["Index2"] and index2_cycle == f"N{self.run_index2_cycles}":
+            return
+        if (
+            self.run_index2_cycles == len(self.sample["Index2"])
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/validator.py]: fixed

File: cg/apps/demultiplex/sample_sheet/validator.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+        raise OverrideCyclesError(message)
+
+    def validate_sample(
+        self,
+        sample: dict[str, str],
+    ) -> None:
+        """Determine if the override cycles are valid for a given sample."""
+        self.sample = sample
+        self.sample_cycles = sample["OverrideCycles"].split(";")
+        self.sample_id = sample["Sample_ID"]
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/validator.py]: fixed

File: tests/fixture_plugins/demultiplex_fixtures/path_fixtures.py [hunk 5]
```diff
@@ -10,6 +10,8 @@
 from cg.models.flow_cell.flow_cell import FlowCellDirectoryData
 
 CORRECT_SAMPLE_SHEET: str = "CorrectSampleSheet.csv"
+V2_SAMPLE_SHEET: str = "SampleSheet_bcl_convert.csv"
+V1_SAMPLE_SHEET: str = "SampleSheet_bcl2fastq.csv"
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply tests/fixture_plugins/demultiplex_fixtures/path_fixtures.py:14]: turned into fixtures

**Comment by diitaz93** [comment]
@karlnyr I increased this PR's scope slightly and implemented the changes in #2958. Just so you know – the changes you suggested are implemented there. There fore I am closing this PR

**Pull request closed without merging**